controllers/agents/app/servicebinding_controller.go

```diff
@@ -432,12 +441,30 @@ func (r *ServiceBindingReconciler) updateContainerInfo(ctx context.Context, cont
 		}
 
 		for _, e := range sb.Spec.Env {
-			c.Env = append(c.Env, v1.EnvVar{
-				Name:  e.Name,
-				Value: string(psSecret.Data[e.Key]),
-			})
-
+			if checkIfEnvPresentOrNot(c.Env, v1.EnvVar{
+				Name: e.Name,
+				ValueFrom: &v1.EnvVarSource{
+					SecretKeyRef: &v1.SecretKeySelector{
+						Key: e.Key,
+						LocalObjectReference: v1.LocalObjectReference{
+							Name: psSecret.Name,
+						},
+					},
+				},
+			}) {
+				c.Env = append(c.Env, v1.EnvVar{
+					Name: e.Name,
+					ValueFrom: &v1.EnvVarSource{
+						SecretKeyRef: &v1.SecretKeySelector{
+							Key: e.Key,
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: psSecret.Name,
+							},
+						},
+					}})
+			}
```

Comment:
⚠️ this should be superseded by next comment

this can be simplified
```suggestion
			e := v1.EnvVar{
				Name: e.Name,
				ValueFrom: &v1.EnvVarSource{
					SecretKeyRef: &v1.SecretKeySelector{
						Key: e.Key,
						LocalObjectReference: v1.LocalObjectReference{
							Name: psSecret.Name,
						},
					},
				},
			}
			if checkIfEnvPresentOrNot(c.Env, e) {
				c.Env = append(c.Env, e)
			}
```

However, if the EnvVar is already set but not up-to-date, it won't be updated.
I'd propose to introduce a new function named `createOrUpdateEnvVar(c *v1.Container{}, ev v1.EnvVar)` that adds the EnvVar if it does not exist, or update that if it exists.

Reply:
We are only referring the secret in the field. And we do not allow sci to be updated in the service claim. So once the env is created , we won't need to update it's keys. Values would be updated in the secret itself. WDYT?

Reply:
The user may change the secret key referenced by the EnvVar. In this case the change won't be persisted. I'd propose to use the approach described in the next comment for a more solid approach